src/planner/PlanNode.h

```diff
@@ -216,23 +234,57 @@ class SingleInputNode : public SingleDependencyNode {
 public:
     void setInputVar(std::string inputVar) {
         DCHECK(!inputVars_.empty());
-        inputVars_[0] = std::move(inputVar);
+        auto* inputVarPtr = qctx_->symTable()->getVar(inputVar);
+        DCHECK(inputVarPtr != nullptr);
+        std::string oldVar;
+        if (inputVars_[0] != nullptr) {
+            oldVar = inputVars_[0]->name;
+        }
+        inputVars_[0] = inputVarPtr;
+        if (!oldVar.empty()) {
+            for (auto& output : outputVars_) {
+                qctx_->symTable()->deleteDerivative(oldVar, output->name);
+                qctx_->symTable()->addDerivative(inputVar, output->name);
+                qctx_->symTable()->deleteDependency(output->name, oldVar);
+                qctx_->symTable()->addDependency(output->name, inputVar);
+            }
+            qctx_->symTable()->deleteRead(oldVar, this);
+            qctx_->symTable()->readBy(inputVar, this);
+        } else {
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(inputVar, output->name);
+                qctx_->symTable()->addDependency(output->name, inputVar);
+            }
+            qctx_->symTable()->readBy(inputVar, this);
```

Comment:
Maybe we'd better to extract this logic to a `addInputVar` function so SingleInputNode also could use it.